src/phase0/state_transition/epoch_processing.rs

```diff
@@ -519,6 +539,120 @@ pub fn process_participation_record_updates<
+    /*
+    Return the combined effective balance of the set of unslashed validators participating in ``attestations``.
+    Note: ``get_total_balance`` returns ``EFFECTIVE_BALANCE_INCREMENT`` Gwei minimum to avoid divisions by zero.
+     */
+    get_total_balance(
+        state,
+        &get_unslashed_attesting_indices(state, attestations, context)?,
+        context,
+    )
+}
+
+pub fn weigh_justification_and_finalization<
+    const SLOTS_PER_HISTORICAL_ROOT: usize,
+    const HISTORICAL_ROOTS_LIMIT: usize,
+    const ETH1_DATA_VOTES_BOUND: usize,
+    const VALIDATOR_REGISTRY_LIMIT: usize,
+    const EPOCHS_PER_HISTORICAL_VECTOR: usize,
+    const EPOCHS_PER_SLASHINGS_VECTOR: usize,
+    const MAX_VALIDATORS_PER_COMMITTEE: usize,
+    const PENDING_ATTESTATIONS_BOUND: usize,
+>(
+    state: &mut BeaconState<
+        SLOTS_PER_HISTORICAL_ROOT,
+        HISTORICAL_ROOTS_LIMIT,
+        ETH1_DATA_VOTES_BOUND,
+        VALIDATOR_REGISTRY_LIMIT,
+        EPOCHS_PER_HISTORICAL_VECTOR,
+        EPOCHS_PER_SLASHINGS_VECTOR,
+        MAX_VALIDATORS_PER_COMMITTEE,
+        PENDING_ATTESTATIONS_BOUND,
+    >,
+    total_active_balance: Gwei,
+    previous_epoch_target_balance: Gwei,
+    current_epoch_target_balance: Gwei,
+    context: &Context,
+) -> Result<(), Error> {
+    let previous_epoch = get_current_epoch(state, context);
+    let current_epoch = get_current_epoch(state, context);
+    let old_previous_justified_checkpoint = state.previous_justified_checkpoint.clone();
+    let old_current_justified_checkpoint = state.current_justified_checkpoint.clone();
+
+    // Process justifications
+    state.previous_justified_checkpoint = state.current_justified_checkpoint.clone();
+    for i in (1..(JUSTIFICATION_BITS_LENGTH - 1)).rev() {
+        let prev = state
+            .justification_bits
+            .get(i - 1)
+            .expect("at least one bit in justification_bits");
+        state.justification_bits.set(i, prev);
+    }
+    state.justification_bits.set(0, false);
+    if previous_epoch_target_balance * 3 >= total_active_balance * 2 {
+        state.current_justified_checkpoint = Checkpoint {
+            epoch: previous_epoch,
+            root: *get_block_root(state, previous_epoch, context)
+                .expect("block root exists for checkpoint"),
+        };
+        state.justification_bits.set(1, true);
+    }
+    if current_epoch_target_balance * 3 >= total_active_balance * 2 {
+        state.current_justified_checkpoint = Checkpoint {
+            epoch: current_epoch,
+            root: *get_block_root(state, current_epoch, context)
+                .expect("block root exists for checkpoint"),
+        };
+        state.justification_bits.set(0, true);
+    }
+
+    // Process finalizations
+    let bits = &state.justification_bits;
+    // The 2nd/3rd/4th most recent epochs are justified, the 2nd using the 4th as source
+    if bits[1..4].all() && old_previous_justified_checkpoint.epoch + 3 == current_epoch {
```

Comment:
With `JUSTIFICATION_BITS_LENGTH` being a constant, I was wondering if the hardcoded `bits[1..4]` is supposed to be used vs. `bits[1..JUSTIFICATION_BITS_LENGTH]`

Reply:
ah good catch, this could use the constant, but it doesn't :)

there is a bit more symmetry in the code w/ digits here, rather than working in offsets from the constant